artiq/coredevice/phaser.py

```diff
@@ -468,6 +474,32 @@ def get_crc_err(self):
         """
         return self.read8(PHASER_ADDR_CRC_ERR)
 
+    @kernel
+    def measure_frame_timestamp(self):
+        """Perform a register read and record the exact frame timing in `self.frame_tstamp`.
+
+        Deterministic timing requires updates to be schedule at multiples of `self.t_frame` later.
+        See `get_next_frame_timestamp()`.
+        """
+        rtio_output((self.channel_base << 8) | (PHASER_ADDR_BOARD_ID & 0x7f), 0)  # can read any register
+        delay_mu(int64(self.t_frame))
+        self.frame_tstamp = rtio_input_timestamp(rtio_get_counter() + 0xffffff, self.channel_base)
+        delay(10*ms)
+
+    @kernel
+    def get_next_frame_timestamp(self, after_timestamp_mu = int64(-1)):
+        """Return the RTIO timestamp of the next frame after `after_timestamp_mu`.
+
+        If `after_timestamp_mu < 0`, return the next frame after `now_mu()`.
+
+        Updates scheduled at this timestamp and multiples of `self.t_frame` later will
+        have deterministic latency with respect to the RTIO timeline.
```

Comment:
What is "latency with respect to the RTIO timeline"?
Better "latency to output".

Reply:
Fixed in 075cb26